### ISSUE TITLE
Conditional branch optimizations

### ISSUE THREAD

**Issue by Itay2805** [vnmakarov/mir#264]
It seems that on the current branch, a conditional branch won't be eliminated even if it was done before, an example:

```
test:	module
test:	func	p, p:test

# 1 arg, 0 locals
	bf	L1, test
	ret	1
L1:
	bf	L2, test
	ret	1
L2:
	ret	0
	endfunc
	endmodule
``` 

gets converted into 

```x86asm
mov    %rdi,%rax
cmp    $0x0,%rax
je     7f7d921231b9 <code+0x19>
mov    $0x1,%rax
jmp    7f7d921231d1 <code+0x31>
cmp    $0x0,%rax
je     7f7d921231cf <code+0x2f>
mov    $0x1,%rax
jmp    7f7d921231d1 <code+0x31>
xor    %eax,%eax
ret
```

Given how dynamic languages have alot of implicit checks it might be worth it to have such an optimization (and I assumed it would already happen in `Sparse Conditional Constant Propagation` but that is just from the name, not from actual knowledge about how a jit works)

It might even be nice to use the block versioning you talked about to have optimizations/de-optimizations applied based on such checks (which btw I am excited to see the new property and aliasing stuff get documented property and start playing with them).

**Comment by vnmakarov** [comment]
Thank you for the test.  I am really interesting in them.

For simplicity of epilogue generation (stack de-allocation and restoring used callee-saved regs), MIR is transformed to have one place for return.  In given case epilogue is just one return and we can have them in several places.

So I think the generated case can be improved by a small pass after prologue which changes jmp to ret just to ret.

Also I have jump opt pass on bbv branch and I guess I can add there case bf L1, test; ... bf L2, test => bf L2, test.

Btw, I am interesting in implementation of builtin_expect in c2m.  It will permit to make frequently executed code strait, removing number for executing jumps and improving code locality.  That is what I see on test cases of JIT I am working now.

**Comment by Itay2805** [comment]
In general it would be interesting to pass information such as conditional result in the constant propogation for example in this case we know the register can't be zero if we continue after the first check, making it so we can remove the second check/turn it into a unconditional branch. This can be especially useful later on for eliminating range checks for array accesses and so on.

**Comment by Itay2805** [comment]
yeah, I think a constant value check elimination would go a long way as is, especially with alot of managed languages needing to insert NULL checks everywhere in the code.
